### PR TITLE
Modifies cookie cache value check to also check with redis-based value

### PR DESCRIPTION
**Pull request by lankyfrenchman**
Partially addresses RadiusNetworks/Iris#1299

If an app includes this repo, and that app has an instance of Redis defined to track user sessions, these changes will check with Redis that the session is valid and hasn't been ended. This addresses an issue where a cookie may be reused maliciously, since although we clear cookies on logout, if the cookie has been retained it could be put into place elsewhere and still be valid.

Other PRs that will use this:
https://github.com/RadiusNetworks/kracken/pull/525
https://github.com/RadiusNetworks/iris/pull/1828

### CODE REVIEW

File: lib/kracken/controllers/authenticatable.rb [hunk 1]
```diff
@@ -124,6 +112,30 @@ def user_signed_in?
 
       private
 
+      def session_present?
+        session[:user_id] && session[:user_cache_key]
```
Comment (cupakromer) [inline lib/kracken/controllers/authenticatable.rb:116]: Can we add a note stating we are intentionally not checking `session[:user_uid]` here to handle the existing session upgrade case? Or am I miss understanding this check?

File: lib/kracken/controllers/authenticatable.rb [hunk 2]
```diff
@@ -124,6 +112,30 @@ def user_signed_in?
...
+      def session_and_redis_match?
+        ::SessionManager.get(session[:user_uid]) == session[:user_cache_key]
+      end
+
+      def delete_session_data
+        # Sign out current user
+        session.delete :user_id
+
+        # Clear that user's cache key
+        session.delete :user_cache_key
```
Comment (cupakromer) [inline lib/kracken/controllers/authenticatable.rb:128]: I think we're missing the `user_uid` here:

```suggestion
        # Clear that user's cache data
        session.delete :user_uid
        session.delete :user_cache_key
```

File: lib/kracken/controllers/authenticatable.rb [hunk 3]
```diff
@@ -124,6 +112,30 @@ def user_signed_in?
...
+      end
+
+      def clear_cache_cookie_and_sign_out
+        # Delete the cookie to prevent redirect loops
+        cookies.delete :_radius_user_cache_key
+
+        # Redirect to the account app
+        redirect_to_sign_in
+      end
```
Comment (cupakromer) [inline lib/kracken/controllers/authenticatable.rb:137]: Am I missing something or is this not being used?

File: lib/kracken/controllers/authenticatable.rb [hunk 4]
```diff
@@ -71,25 +71,13 @@ def check_token_expiry!
       #    delete the cookie
       #
       def handle_user_cache_cookie!
-        if cookies[:_radius_user_cache_key]
-          if cookies[:_radius_user_cache_key] == "none"
-            # Sign out current user
-            session.delete :user_id
+        return redirect_to_sign_in unless session_present?
```
Comment (cupakromer) [inline lib/kracken/controllers/authenticatable.rb:74]: Ok, I think I understand why the original code had the wrapping `if cookies[:_radius_user_cache_key]` guard. Not all of our applications require login on every page. And when those pages are setup we are only using `skip_before_actoin :authenticate_user!` on them. The idea being, if there is a user session then we'll handle a cache mismatch, otherwise we'll proceed and let the `authenticated_user!` handle checking if they are signed in and redirecting if they are not.

Basically, that's a lot of words to say, I think this just needs to be:

```suggestion
        return unless session_present?
```

**Comment by cupakromer** [comment]
Looks like everything's been addressed except the `check_token_expiry!` which still needs to be updated.

**Comment by lankyfrenchman** [comment]
This should be good to go now. I'm not sure what's up with CI though.

**Comment by cupakromer** [comment]
It should fail for you locally as well. It's failing because the changes here changed that part of the code. The specs need to be updated to no-longer rely on the legacy user cache key `cookie`.

**Comment by cupakromer** [comment]
Looks like this needs to rebase against the main branch to get the recent changes to the CI process.

**Comment by lankyfrenchman** [comment]
@cupakromer updated some tests, rebased, and CI is now happy.
